src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/EditorConfigFileWatcher.cs

```diff
@@ -0,0 +1,271 @@
+                    _treeWatcher = ProjectDataSources.SyncLinkTo(treeSource, editorconfigItemsSource, target);
+
+                    return Task.CompletedTask;
+                });
+        }
+
+        private CancellationTokenSource CreateLinkedCancellationTokenSource()
+        {
+            // we want to cancel when we switch what file is watched, or when the project is unloaded
+            if (_projectServices.Project?.Services?.ProjectAsynchronousTasks?.UnloadCancellationToken != null)
+            {
+                return CancellationTokenSource.CreateLinkedTokenSource(
+                    _projectServices.Project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+            }
+            else
+            {
+                return new CancellationTokenSource();
+            }
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                _taskDelaySchedulerCancellationTokenSource.Cancel();
+                _taskDelaySchedulerCancellationTokenSource.Dispose();
+                _taskDelayScheduler.Dispose();
+                _treeWatcher.Dispose();
+
+                var filesBeingWatched = _filesBeingWatched.Keys.ToArray();
+                await UnregisterFileWatchersAsync(filesBeingWatched);
+            }
+        }
+
+        private static SortedSet<string> GetEditorConfigFilePaths(IProjectTree newTree, IProjectSubscriptionUpdate projectUpdate)
+        {
+            var editorConfigFilePaths = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);
+            string projectFilePath = projectUpdate.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildProjectFullPathProperty, defaultValue: null);
+            if (string.IsNullOrEmpty(projectFilePath))
+            {
+                return editorConfigFilePaths;
+            }
+
+            // TODO (tmeschter): if $(DiscoverEditorConfigFiles) is "false", disable all file watchers by returning an empty set.
+
+            string projectDirectory = Path.GetDirectoryName(projectFilePath);
+            string directory = Path.GetDirectoryName(projectDirectory);
+            while (directory != null)
+            {
+                string editorConfigFilePath = Path.Combine(directory, ".editorconfig");
+                editorConfigFilePaths.Add(editorConfigFilePath);
+            }
+
+            return editorConfigFilePaths;
+        }
+
+        private async Task HandleFileChangedAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                TraceUtilities.TraceVerbose("A .editorconfig file has been added or removed. Marking project dirty.");
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+#pragma warning disable RS0030
+                await _projectServices.Project.Services.ProjectAsynchronousTasks.LoadedProjectAsync(async () =>
+#pragma warning restore RS0030
```

Comment:
When one of the .editorconfig files outside the project cone has changed we need to re-evaluate/re-run the design-time build. What should I use instead of this banned API?

Reply:
use `_projectTasksService`

---
In reply to: [260990526](https://github.com/dotnet/project-system/pull/4619#discussion_r260990526) [](ancestors = 260990526)

Reply:
If the warning didn't provide enough info, can you please PR a better message?